python/lsst/daf/butler/datastores/fileDatastore.py

```diff
@@ -1762,6 +1763,105 @@ def getURI(self, ref: DatasetRef, predict: bool = False) -> ResourcePath:
             )
         return primary
 
+    def getManyURIs(
+        self,
+        refs: Iterable[DatasetRef],
+        predict: bool = False,
+        throwIfNotFound: bool = True,
+    ) -> Dict[DatasetRef, DatasetRefURIs]:
+        # Docstring inherited
+
+        PrimaryType = Optional[ResourcePath]
+        ComponentsType = Dict[str, ResourcePath]
+        uris: Dict[DatasetRef, DatasetRefURIs] = {}
+
+        records = self._get_stored_records_associated_with_refs(refs)
+
+        missingRefs = [ref for ref in refs if ref.id not in records.keys()]
```

Comment:
`records` is just a `dict` so the result of `records.keys()` is a `dict_keys`, which is designed to be iterated over and behaves something like a generator if I understand it correctly.

For that matter if this was going to be unchanged I probably should have changed the whole statement from a list comprehension to a generator IE
```
missingRefs = (ref for ref in refs if ref.id not in records.keys())
```
because there's no need to materialize the whole `missingRefs` list (I did think it was helpful to name the list though, instead of using it directly in line 1782).

But, as you note on line 1818 maybe we shouldn't iterate over the whole list of refs again, so this will want some rewriting to keep track of the not-missing refs.

Reply:
I think `records.keys()` is set-like so `in` should be fast. I was mainly wondering whether there is a cost to python to create it or if it's a read-only view of something that is free to return each time.

Reply:
oh, I see. Yes, it's probably reasonable & best to move it out of the comprehension & cache it.